docs/topics/wait_for.rst

```diff
@@ -0,0 +1,86 @@
+
+Checks and Timeouts
+~~~~~~~~~~~~~~~~~~~~
+Whilst the above example shows off waiting for a message, this will return as soon as *any* message is sent.
+Oftentimes, we want to narrow it down to a specific message, which we can do by passing a predicate to the ``check`` keyword-argument
+
+.. code-block:: python3
+   :emphasize-lines: 2, 3, 4
+
+    await message.channel.send('say hi!')
+
+    def check(m: discord.Message):
+        # check that the author is the one who sent the original message
+        return m.author == message.author 
+
+    msg: discord.Message = await client.wait_for('message', check=check)
+    await msg.reply('hello!')
+
+In this example, ``wait_for`` will only accept the incoming message when the predicate returns ``True``.
+The arguments passed to the predicate match the signature of the corresponding event.
+For example, when defining a message event handler we do `async def on_message(message)` - a predicate would similarly be defined with `def check(message)`.
+
+Likewise, a predicate for `reaction_add` would take `def check(reaction, user)`.
+
+It's common to also include a timeout in addition to a check so our code doesn't potentially end up waiting indefinitely, or to add a limited time window for the author to send a message.
+.. code-block:: python3
+   :emphasize-lines: 2, 3, 4, 5
+
+    await message.channel.send('say hi!')
+    try:
+        msg: discord.Message = await client.wait_for('message', timeout=5.0) # timeout in seconds.
+    except asyncio.TimeoutError: 
+        await message.channel.send('You did not respond :(')
+    else:
+        await msg.reply('hello!')
+
+We pass the timeout in seconds to the ``timeout`` keyword-argument. Timeouts behave like :meth:`asyncio.wait_for`, so we do the usual `try` and `except`, catching :class:`asyncio.TimeoutError`:
+.. warning::
+
+    Avoid calling :meth:`Client.wait_for` within a loop to catch events. Due to the nature of async IO, events may be fired between loops, causing the loop to miss events.
+
+Examples
+~~~~~~~~~~~~~~~~~~~~
+Wait for reaction
++++++++++++++++++
+
+.. code-block:: python3
+
+    def check(reaction: discord.Reaction, user: discord.User):
+        return user == message.author and reaction.message == message # check that the reaction is on a specific message and by a specific user
+
+    reaction, user = await client.wait_for('reaction_add', check=check)
+    await message.channel.send(f'You reacted with {reaction.emoji}!')
+
+Notice the ``reaction_add`` event, unlike the ``message`` event, takes 2 arguments. Thus, the check function takes the same arguments as that, ``reaction, user``.
+
+
+Closing Remarks
+~~~~~~~~~~~~~~~~~~~~
+``wait_for`` is a powerful tool used often to wait for responses in code. The examples above show only 2 types of ``wait_for``, reactions and messages, but you can wait for any event! A full list of events can be seen here: :ref:`event reference <discord-api-events>`. 
```

Comment:
While they are powerful I think that it should be mentioned that Buttons and Modals can be also used for similar functionality that provide better UX for users instead and could/should be considered of waiting for a message or reaction. 

However I can also understand that this may be out of scope for the `wait_for` guide and probably better to be added to the views/modal guide instead.

Also I am a bit struggling to find good use cases on waiting for other events than reaction or `on_message` events, especially ones that aren't superseded by UI components. Because I think the examples should show some functionality that is not possible with those, however again that is just a major personal nitpick mostly but I wanted to mention it.

Only one example I can think of could be having a `wait_for` for the `voice_state_update` event to join with a user who invoked a command that needs someone to be in voice before doing something. Roughly working like so:

1. Command should play something in the voice channel the user is in
2. check if user is in voice if yes join and play sound
3. else send a message asking them to join a voice channel
4. wait_for user to join a voice channel with timeout
5. play sound / file after user joined.

It is a bit convoluted but also a bit more of an involved example that could show off the capability of `wait_for`